Improve error message for lookup in function types

mestorlx opened the pull request:
Fixes #4066 

A new function was added in TypeChecker to improve the following message "member not found after argument-dependant lookup in type ..." when type is a function. 

when a.b.test() is called the compiler will suggest a.b().test()

erak commented on Changelog.md:
```diff
@@ -116,6 +116,7 @@ Bugfixes:
  * Type Checker: Fix freeze for negative fixed-point literals very close to ``0``, such as ``-1e-100``.
  * Type Checker: Dynamic types as key for public mappings return error instead of assertion fail.
  * Type Checker: Fix internal error when array index value is too large.
+ * Type Checker: Search for a name suggestion when the name is not found after argument-dependant lookup.
```
There's a tiny typo. Should be ``argument-dependent``

mestorlx replied:
Fixed. Thanks

erak commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2514,3 +2519,40 @@ void TypeChecker::requireLValue(Expression const& _expression)
+
+	for (ASTPointer<ASTNode> const &node : _contract.subNodes())
+	{
+		auto decl = dynamic_cast<Declaration const*>(&(*node));
```
Please avoid using the ``NULL`` macro (use ``nullptr`` if necessary). Also this can be one line:
```
if (auto const* decl = dynamic_cast<Declaration const*>(...))
{ ... }
```

mestorlx replied:
changed

erak commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2514,3 +2519,40 @@ void TypeChecker::requireLValue(Expression const& _expression)
+				ContractDefinition const& contract = contractType->contractDefinition();
+				if (getNameSuggestion(contract, _missingFunctionName, _suggestion,false))
+				{
+					if(_first==false)
```
By swapping the ``if`` and ``else`` case, you could write ``if (_first)``.

mestorlx replied:
changed

erak commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2514,3 +2519,40 @@ void TypeChecker::requireLValue(Expression const& _expression)
+		if (decl != NULL)
+		{
+			auto contractType = dynamic_pointer_cast<ContractType const>(decl->type());
+			if (contractType != NULL)
```
Same as above.

mestorlx replied:
changed

axic commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2514,3 +2519,40 @@ void TypeChecker::requireLValue(Expression const& _expression)
+)
+{
+
+	for (ASTPointer<ASTNode> const &node : _contract.subNodes())
```
I think indentation in this file is slightly off.

mestorlx replied:
I see it ok in my editor (Visual Studio Code)

erak commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2142,9 +2142,14 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 		}
+
+		string functionNameSuggestion =".";
+		if(getNameSuggestion(*m_scope, memberName, functionNameSuggestion))
```
Instead of passing the ``functionNameSuggestion`` by non-const reference, setting it in the function and then using a ``bool`` to signal if there's a suggestion or not, ``getNameSuggestion`` could return the suggestion right away or an empty string if there's non.

mestorlx replied:
for this change I would need to do something like this.
```
string suggestion = getNameSuggestion(contract, _missingFunctionName,false);
if (!suggestion.empty())
  {
  if (_first)
     suggestion = decl->name() + "." + suggestion;
  else
    suggestion = decl->name() + "()." + _suggestion;
  return suggestion;
 }
```

It make sense to make this change?

mestorlx replied:
@chriseth Should I implement this change?

axic commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2142,9 +2142,14 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 		}
+
+		string functionNameSuggestion =".";
+		if(getNameSuggestion(*m_scope, memberName, functionNameSuggestion))
```
Please keep a space between `if` and `(`

mestorlx replied:
changed

axic commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2142,9 +2142,14 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 				);
 		}
+
+		string functionNameSuggestion =".";
```
Space after the equal sign.

mestorlx replied:
changed

axic commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2514,3 +2519,40 @@ void TypeChecker::requireLValue(Expression const& _expression)
+	for (ASTPointer<ASTNode> const &node : _contract.subNodes())
+	{
+		auto decl = dynamic_cast<Declaration const*>(&(*node));
+		if (decl != NULL)
```
Could merge these into a single if statement:
```
if (auto declaration = dynamic_cast<Declaration const*>(&(*node)))`
```

Though are you sure this is the best cast possible? :)

mestorlx replied:
changed. Thanks

erak reviewed (changes requested):
Besides the comments about simplification of the code, the coding style is a little bit off. Please add a whitespace to ``if ()`` statements, assignments and before each parameter in function calls.

erak commented:
Also needs a rebase.

erak commented:
@mestorlx Did something go wrong? This PR contains a lot of commits that are already on ``solidity/develop``. Can you try to clean this up? We can also help you with figuring this out.

mestorlx commented:
> @mestorlx Did something went wrong? This PR contains a lot of commits that are already on `solidity/develop`. Can you try to clean this up? We can also help you with figuring this out.

@bit-shift Yes when I rebased I applied a lot of commits, sorry about that. 
Fixed now, Thanks @chriseth !

mestorlx commented:
> > @mestorlx Did something went wrong? This PR contains a lot of commits that are already on `solidity/develop`. Can you try to clean this up? We can also help you with figuring this out.
> 
> @bit-shift Yes when I rebased I applied a lot of commits, sorry about that. I reverted the commits so the files changed only contain my changes. Not sure how to clean the commits.

chriseth commented:
I don't think this solves the mentioned issue. It should behave as follows:

```
contract c {
function f(uint) internal pure {}
function f(bytes) internal pure {}
function g() {
this.f;
}
}
```

The compiler cannot determine which if the two functions you mean because you did not actually call the function. The error should report: "did you intend to call the function?" (without providing the name because that is obvious, you already gave it in the source).

mestorlx commented:
> I don't think this solves the mentioned issue. It should behave as follows:
> 
> ```
> contract c {
> function f(uint) internal pure {}
> function f(bytes) internal pure {}
> function g() {
> this.f;
> }
> }
> ```
> The compiler cannot determine which if the two functions you mean because you did not actually call the function. The error should report: "did you intend to call the function?" (without providing the name because that is obvious, you already gave it in the source).

I understand the point you are making. 
In the case that was mentioned in issue #4060 the behaviour should be the same?
I should remove the function name suggestion and just leave the message "Did you intend to call the function."?

chriseth commented:
Yes, in the issue, where it had `a.b.test()` it should also provide the same error message - no need to mention the name (although you can, but it should be unique).

mestorlx commented:
Great, changed.
What tests should I run to verify its not producing undesired output. 
I can check that the message is shown when it should (with examples) not so sure how to check the message is not shown in other places.

chriseth reviewed (changes requested):
Please start with writing a test, because I think you are not approaching the problem from the right angle. Note that the error message ("did you intend to call the function") should not be displayed when the function has actually been called. Also searching the current contract for functions will not get you very far if the function to be called is taken from another contract.

chriseth commented:
The test 576_member_getter_call_without_parentheses.sol is correct, the other one is not - calling the function will not help.

So the problem basically is as follows: If we search for a member in a function type and it is not present, then the user probably wanted to call the function and select the member in the return value of the function.

mestorlx commented:
Added more tests and removed the one that was not ok.

chriseth commented:
Do you know how to do this now? If you have a member access of a function type, ask the user if they intended to call the function.

mestorlx commented:
@chriseth. I think I do.
Current tests (and code) cover getter functions. It this ok? 
If I extend to other methods like below, should I do it here or on a new PR?

```
pragma solidity ^0.4.0;

contract B{
   function f() public returns (bool){
       return true;
   }
}
contract C{
    B b;
    function g() public returns (bool){
        return b.f;
    }
}
```

chriseth commented:
Please read my comment above again. The code you are writing should not perform name lookup and it should not have to care whether it is a getter or not.

mestorlx commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2140,9 +2140,14 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 		}
+
+		string functionNameSuggestion = ".";
+		if (exprType->category() == Type::Category::Function)
```
@chriseth just changed the approach. 
I need to add extra checks to verify that the member is visible and fix other tests that are failing. 
Is this the correct approach ?

chriseth replied:
Yes, this is the right approach.
Note that this PR should be tiny. No need to check visibility or computing a name suggestion.

Combine the type check here with the `memberName == "value"` check below to generate the correct second part of the error message and you are done.

Commit:
Improved error message for lookup in function types.

mestorlx commented:
Hi @chriseth  I added the change suggested. I needed to add extra code to prevent the message being added when gas (syntaxTests/nameAndTypeResolution/495_builtin_ripemd160_reject_gas.sol) or selector (syntaxTests/nameAndTypeResolution/485_function_types_selector_2.sol) was being called. Also  to prevent the message to be added when a declaration shadows the class (syntaxTests/constructor/overriding_constructor.sol).

I have some doubts regarding the behavior, for instance in the following example the code will not compile:

```
    contract A{
    }
    contract B{
        A public a;
    }
    contract C{
        B b;
        function f() public view{
            b.a.f();
        }
    }
```
And we are showing did you intend to call the function. Here if we call the function (as below) the code will not compile, however we will not show "did you intend to call the function"
```
    contract A{
    }
    contract B{
        A public a;
    }
    contract C{
        B b;
        function f() public view{
            b.a().f();
        }
    }
```
I believe this is not part of this PR but maybe is something that could be addressed in a different PR?

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2140,9 +2140,27 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 		string errorMsg = "Member \"" + memberName + "\" not found or not visible "
 				"after argument-dependent lookup in " + exprType->toString() +
-				(memberName == "value" ? " - did you forget the \"payable\" modifier?" : ".");
+				(memberName == "value" ? " - did you forget the \"payable\" modifier?" :
```
Please combine this with what you did above.

chriseth commented:
Can you provide an example why you need to check if the function returns a contract or struct?

mestorlx commented:
> Can you provide an example why you need to check if the function returns a contract or struct?

If I just add `if (exprType->category() == Type::Category::Function)` then for the following cases the message "Did you intend to call the function?" will be added (wrongly) to the following cases.

```
    contract A {
        function f() public {
            keccak256.gas();
        }
    }
```
For this case alternatively I could check that name is not gas.

```
    contract B{
        function g() pure internal {
        }
        function f() public view returns (bytes4) {
            return g.selector;
        }
    }
```
again here I could alternatively check that the function name is not selector.

```
    contract C { function f() public {} }
    contract D is C {
      function C() public pure returns (uint8) {}
      function g() public {
        C.f();
      }
    }
```
For this case checking the struct or class will ensure that the "Did you intend to call the function?" will not be appended since its calling `function C() public pure returns (uint8)` rather than the getter.

chriseth commented:
Ah that's very clever! If possible, a more dynamic approach would be to check that the function if called, would return exactly one value and call `t->members(m_scope).membersByName(memberName)` on that type to check if the member is there.

chriseth commented:
This has the benefit that if we add more types in the future, we would not have to change this code.

mestorlx commented:
@chriseth thanks for the feedback.
 I can certainly make that change. However in the following contract
```
    contract A{
        function f() private pure{

        }
    }
    contract B{
        A public a;
    }
    contract C{
        B b;
        function f() public view{
            b.a.f();
        }
    }
```

The member would not be there, hence the "did you intend to call the function" will not be appended. This is maybe a particular problem since there are two problems, calling a getter with out `()` and calling a private member from outside the contract. Clearly the user intended to call `  b.a().f();` i.e the getter. However, calling the getter will not help because `function f() private pure` is private and not accessible. The same would happened if f is not define. 
I think the question here is more conceptual. In this case should we or should we not show "did you intend to call the function?".

chriseth commented:
@mestorlx you are right, perhaps your simple solution is the better one. Please still combine it with the case about ".value".

mestorlx commented:
@chriseth updated.

chriseth commented on Changelog.md:
```diff
@@ -73,6 +73,7 @@ Breaking Changes:
  * Syntax Checker: Strictly require visibility specifier for functions. This was already the case in the experimental 0.5.0 mode.
  * Syntax Checker: Disallow unary ``+``. This was already the case in the experimental 0.5.0 mode.
  * Syntax Checker: Disallow single statement variable declaration inside if/while/for bodies that are not blocks.
+ * Syntax Checker: Improved error message for lookup in function types.
```
This is not a breaking change. Please move it to the "compiler features" section.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2141,8 +2141,25 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+		{
+			if (auto funType = dynamic_pointer_cast<FunctionType const>(exprType))
+			{
+				auto t = funType->returnParameterTypes();
```
Please use `auto const&`.

chriseth reviewed (changes requested):
Great, almost there! Please also rebase and squash into a single commit.

mestorlx commented:
@chriseth done.
hopefully ready to merge now.

chriseth commented:
Thanks a lot!
